File: app/src/main/java/yagoc/pieces/Queen.java
package yagoc.pieces;

import yagoc.board.BoardView;
import yagoc.board.Move;
import yagoc.board.Square;

import java.util.stream.Stream;

import static yagoc.pieces.Bishop.generateMovesForBishop;
import static yagoc.pieces.Bishop.isCorrectMoveForBishop;
import static yagoc.pieces.Rook.generateMovesForRook;
import static yagoc.pieces.Rook.isCorrectMoveForRook;

public class Queen extends Piece {
    public Queen(PieceColor pieceColor) {
        super(PieceType.Queen, pieceColor);
    }

    @Override
    public boolean isValidForPiece(BoardView board, Move move) {
        return (isCorrectMoveForBishop(board, move) || isCorrectMoveForRook(board, move));
    }

    @Override
    public Stream<Move> generateMovesForPiece(BoardView board, Square from) {
        return Stream.concat(generateMovesForBishop(board, from), generateMovesForRook(board, from));
    }
}

File: app/src/main/kotlin/yagoc/pieces/Queen.kt
package yagoc.pieces

import yagoc.board.BoardView
import yagoc.board.Move
import yagoc.board.Square
import yagoc.pieces.Rook
import java.util.stream.Stream

class Queen(pieceColor: PieceColor) : Piece(PieceType.Queen, pieceColor) {
    public override fun isValidForPiece(board: BoardView, move: Move): Boolean {
        return Bishop.isCorrectMoveForBishop(board, move) || Rook.Companion.isCorrectMoveForRook(board, move)
    }

    public override fun generateMovesForPiece(board: BoardView, from: Square): Stream<Move> {
        return Stream.concat(
            Bishop.generateMovesForBishop(board, from),
            Rook.generateMovesForRook(board, from)
        )
    }
}